asterius/src/Asterius/Types/SymbolMap.hs

```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Module      :  Asterius.Types.SymbolMap
+-- Copyright   :  (c) 2018 EURL Tweag
+-- License     :  All rights reserved (see LICENCE file in the distribution).
+--
+-- The @'SymbolMap' a@ type represents a finite map/dictionary from keys of
+-- type 'EntitySymbol' to values of type @a@. Internally it is represented as
+-- an 'IM.IntMap' of , where the key of the unique of the 'EntitySymbol' is used
+-- for the indexing. In order to be able to access the 'EntitySymbol'
+-- corresponding to each entry, we also store it alongside each element.
+module Asterius.Types.SymbolMap
+  ( -- * SymbolMap type
+    SymbolMap,
+
+    -- * Construction
+    empty,
+    singleton,
+
+    -- * Query
+    member,
+    lookup,
+    (!),
+
+    -- * Size
+    size,
+
+    -- * Insertion
+    insert,
+
+    -- * Filtering
+    filterWithKey,
+    filter,
+    restrictKeys,
+
+    -- * Folds and Maps
+    foldrWithKey',
+    mapWithKey,
+    mapAccum,
+    mapKeys,
+
+    -- * Conversion
+    elems,
+    keys,
+    keysSet,
+
+    -- ** Lists
+    toList,
+    fromList,
+
+    -- ** Maps
+    toMap,
+    fromMap,
+  )
+where
+
+import Asterius.Binary.Orphans ()
+import Asterius.Types.EntitySymbol
+import Binary
+import Control.Monad
+import Data.Data
+import qualified Data.IntMap as IM
+import qualified Data.IntSet as IS
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import GHC.Exts (IsList (..))
+import GHC.Stack
+import Unique
+import Prelude hiding (filter, lookup)
+
+-- | A map from 'EntitySymbol's to values @a@.
+newtype SymbolMap a = SymbolMap (IM.IntMap (EntitySymbol, a))
+  deriving newtype (Eq)
+  deriving stock (Data)
+
+instance Show a => Show (SymbolMap a) where
+  showsPrec d m =
+    showParen (d > 10) $
+      showString "fromList " . shows (toListSM m)
+
+instance Semigroup (SymbolMap a) where
```

Comment:
The `Semigroup`/`Monoid` instances also belong to `deriving newtype`.

Reply:
Done.